### PR TITLE
Add support for Patch OBB

### PR DESCRIPTION
**Pull request by jeevithiesh**

### CODE REVIEW

File: goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java [hunk 1]
```diff
@@ -53,14 +53,18 @@
...
      */
-    public static final XAPKFile xAPK = new
-
-            XAPKFile(
-            true, // true signifies a main file
-            5, // the version of the APK that the file was uploaded
-            // against
-            1721206904L // the length of the file in bytes
-    );
+    public static final XAPKFile[] xAPKS = {
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java]: Make "s" small

File: goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java [hunk 2]
```diff
@@ -53,14 +53,18 @@
...
+        new XAPKFile(
+        true, // true signifies a main file
+        5, // the version of the APK that the file was uploaded
+        1721206904L  // the length of the file in bytes
+        ),
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java]: Press "Ctrl + Alt + L" in all the updated files

File: goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java [hunk 3]
```diff
@@ -275,9 +279,14 @@ public void onClick(View view) {
      * @return true if they are present.
      */
     boolean expansionFilesDelivered() {
-        String fileName = Helpers.getExpansionAPKFileName(this, xAPK.mIsMain, xAPK.mFileVersion);
-        return Helpers.doesFileExist(this, fileName, xAPK.mFileSize, false);
-    }
+        for (XAPKFile xf : xAPKS) {
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java]: Small s

File: goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java [hunk 4]
```diff
@@ -557,9 +547,7 @@ public void onDownloadProgress(DownloadProgressInfo progress) {
         progress.mOverallTotal = progress.mOverallTotal;
         mPB.setMax((int) (progress.mOverallTotal >> 8));
         mPB.setProgress((int) (progress.mOverallProgress >> 8));
-        mProgressPercent.setText(progress.mOverallProgress
-                * 100 /
-                progress.mOverallTotal + "%");
+        mProgressPercent.setText(Long.toString(progress.mOverallProgress * 100 / progress.mOverallTotal) + "%");
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java]: Check whether "Long.toString" is required here

File: goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java [hunk 5]
```diff
@@ -59,7 +60,9 @@ of this software and associated documentation files (the "Software"), to deal
 import java.util.List;
 import java.util.Locale;
 
-import static chimple.DownloadExpansionFile.xAPK;
+import chimple.DownloadExpansionFile;
+
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java]: Remove this empty line

File: goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java [hunk 6]
```diff
@@ -380,16 +383,38 @@ public static void redirectToPlayStore(Activity context, String appUrl) {
...
+            boolean extractionRequired = false;
+            for (DownloadExpansionFile.XAPKFile xf : xAPKS){
+                if((xf.mIsMain && (xf.mFileVersion != mainFileVersion)) || (!xf.mIsMain && (xf.mFileVersion != patchFileVersion))){
+                    extractionRequired = true;
+                    break;
+                }
+            }
+            if(extractionRequired){
+                Intent intent = new Intent(AppActivity.this, SplashScreenActivity.class);
+                startActivity(intent);
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java]: Move this outside the _else_ construct and avoid code repetition

File: goa/proj.android-studio/goa/src/main/java/utils/Zip.java [hunk 7]
```diff
@@ -116,12 +116,11 @@ public void run() {
                 }
             }
         }
+        boolean extraction = isExtractionSuccessful;
         if (isExtractionSuccessful) {
             flagFile = new File(extractPath + ".success.txt");
             flagFile.createNewFile();
-            SharedPreferences.Editor editor = sharedPref.edit();
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/utils/Zip.java:122]: Why did you remove this?

**Review by ankitk20 — CHANGES_REQUESTED** [review]

File: goa/proj.android-studio/goa/src/main/java/utils/Zip.java [hunk 8]
```diff
@@ -12,16 +12,16 @@
...
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
 
-import static chimple.DownloadExpansionFile.xAPK;
 import static org.cocos2dx.cpp.AppActivity.sharedPref;
 
 public class Zip {
 
-    private static int percent = 0;
+    private static int count = 0;
```
Comment (ankitk20) [inline goa/proj.android-studio/goa/src/main/java/utils/Zip.java]: Why is this variable changed?

**Review by ankitk20 — DISMISSED** [review]

**Review by ankitk20 — DISMISSED** [review]

File: goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java [hunk 9]
```diff
@@ -378,119 +388,102 @@ void validateXAPKZipFiles() {
...
-                    zrf = new ZipResourceFile(fileName);
-                    ZipResourceFile.ZipEntryRO[] entries = zrf.getAllEntries();
-                    /**
-                     * First calculate the total compressed length
-                     */
-                    long totalCompressedLength = 0;
-                    for (ZipResourceFile.ZipEntryRO entry : entries) {
-                        totalCompressedLength += entry.mCompressedLength;
+                for (XAPKFile xf : xAPKs) {
+                    String fileName = Helpers.getExpansionAPKFileName(DownloadExpansionFile.this, xf.mIsMain, xf.mFileVersion);
```
Comment (mathewnik90) [inline goa/proj.android-studio/goa/src/main/java/chimple/DownloadExpansionFile.java]: Use better naming conventions

**Review by mathewnik90 — DISMISSED** [review]

File: goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java [hunk 10]
```diff
@@ -291,14 +292,34 @@ public static boolean launchGameNotification() {
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         sharedPref = getSharedPreferences("ExpansionFile", MODE_PRIVATE);
-        String flagFilePath = "/storage/emulated/0/Android/data/" + getPackageName() + "/files/.success.txt";
-        int currentVersionObb = sharedPref.getInt(String.valueOf(R.string.mainFileVersion), 0);
+        String flagFilePath = "/storage/emulated/0/Android/data/com.maq.xprize.chimple.hindi/files/.success.txt";
```
Comment (mathewnik90) [inline goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/AppActivity.java]: Use getPackageName

File: goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/SplashScreenActivity.java [hunk 11]
```diff
@@ -82,29 +95,56 @@ public void toCallApplication() {
...
-            _zip.close();
             toCallApplication();
+        } catch (IOException e) {
+            System.out.println(e);
+        }
+    }
+
+    public int getTotalExpansionFileSize() {
+        int totalExpansionFileSize = 0;
+        ZipFile zipFile;
```
Comment (mathewnik90) [inline goa/proj.android-studio/goa/src/main/java/org/cocos2dx/cpp/SplashScreenActivity.java]: Use better variable names throughout the file

**Review by ankitk20 — DISMISSED** [review]

**Review by mathewnik90 — APPROVED** [review]

**Pull request merged**